Source/Core/VideoCommon/TextureDecoder_Common.cpp

```diff
@@ -546,61 +566,48 @@ static u32 TexDecoder_DecodePixelCMPR(cmprTexelData data, int s, int t)
   int red1 = Convert5To8((data.color1 >> 11) & 0x1F);
   int red2 = Convert5To8((data.color2 >> 11) & 0x1F);
 
+  auto& c = cmprCache[cacheSet].colors;
+  c[0] = MakeRGBA(red1, green1, blue1, 255);
+  c[1] = MakeRGBA(red2, green2, blue2, 255);
+  if (data.color1 > data.color2)
+  {
+    c[2] = MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1), DXTBlend(blue2, blue1), 255);
+    c[3] = MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2), DXTBlend(blue1, blue2), 255);
+  }
+  else
+  {
+    // color[3] is the same as color[2] (average of both colors), but transparent.
+    // This differs from DXT1 where color[3] is transparent black.
+    c[2] = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 255);
+    c[3] = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 0);
+  }
+
+  cmprCache[cacheSet].tag = cacheTag;
+}
+
+u32 TexDecoder_DecodePixelCMPR(cmprTexelData data, int s, int t)
+{
   int ss = s & 3;
   int tt = t & 3;
   int shift = tt * 8 + (6 - (ss << 1));
   int colorSel = (data.colorSel >> shift) & 3;
-  colorSel |= (data.color1 > data.color2) ? 0 : 4;
-
-  // branchless version
-
-  // uint32_t colors[8] = {
-  //   MakeRGBA(red1, green1, blue1, 255),
-  //   MakeRGBA(red2, green2, blue2, 255),
-  //   MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1), DXTBlend(blue2, blue1), 255),
-  //   MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2), DXTBlend(blue1, blue2), 255),
-  //   //altColors ? MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 255)
-  //   //          : MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1), DXTBlend(blue2,
-  //   blue1), 255),
-  //   //altColors ? MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 0)
-  //   //          : MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2), DXTBlend(blue1,
-  //   blue2), 255) MakeRGBA(red1, green1, blue1, 255), MakeRGBA(red2, green2, blue2, 255),
-  //   MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 255),
-  //   MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 0)
-  // };
-  // return colors[colorSel];
-
-  u32 color = 0;
-
-  switch (colorSel)
+
+  // It's quite expensive to derive the four colors of each block.
+  // So we cache then, indexed by the two colors.
+
+  u32 cacheSet = cmprCacheSet(data.color1, data.color2);
+  u32 cacheTag = cmprCacheTag(data.color1, data.color2);
+
+  // our cache is initialized to all zeros, and we don't have space for a valid bit.
+  // This means tag 0 will end up as valid.
+  // A simple solution is to just never hit on tag zero
```

Comment:
An alternative solution which I didn't implement was to simply pre-populate the cache with the correct colours for tag zero 

Tag zero is a nonsense block color anyway, with three solid blacks plus a solid black with 0% alpha

Reply:
Pre-populating it seems like a better solution to me, and wouldn't be that difficult; it'd be something like this:

```C++
static CMPRCacheEntry s_cmpr_cache[1 << CACHE_HASH_BITS] = {{cmprCacheTag(0, 0),
                                                             {
                                                                 MakeRGBA(0, 0, 0, 255),
                                                                 MakeRGBA(0, 0, 0, 255),
                                                                 MakeRGBA(0, 0, 0, 255),
                                                                 MakeRGBA(0, 0, 0, 0),
                                                             }}};
// Pre-populate tag 0 (for input colors of (0, 0)) with the correct value, since we don't have a
// valid bit and thus couldn't tell if tag 0 has a valid value.  Games aren't likely to use this
// though, since it only has 3 solid blacks and a transparent black.
static_assert(cmprCacheSet(0, 0) == 0, "Tag 0 must be at set 0 for above initialization to work");
```

I'm also not 100% sure if the "unlikely to use this" thing is true; wouldn't tag 0 be used whenever a game has a transparent black section of a texture, which doesn't seem particularly uncommon to me?  (Or is there another way to represent transparent black?)

Reply:
Here's a possibility for how to refactor this:

```C++
namespace
{
static constexpr u32 CACHE_HASH_BITS = 10;

class CMPRColorCache
{
private:
  static constexpr u32 GetCacheSet(u16 color1, u16 color2)
  {
    return (u32(color1) * u32(color2) * 0x9E3779B1) >> (32 - CACHE_HASH_BITS);
  }

  static constexpr u32 GetCacheTag(u16 color1, u16 color2)
  {
    return u32(color1) | (u32(color2) << 16);
  }

  static constexpr std::array<u32, 4> ComputeColors(u16 color1, u16 color2)
  {
    int blue1 = Convert5To8(color1 & 0x1F);
    int blue2 = Convert5To8(color2 & 0x1F);
    int green1 = Convert6To8((color1 >> 5) & 0x3F);
    int green2 = Convert6To8((color2 >> 5) & 0x3F);
    int red1 = Convert5To8((color1 >> 11) & 0x1F);
    int red2 = Convert5To8((color2 >> 11) & 0x1F);

    u32 c0 = MakeRGBA(red1, green1, blue1, 255);
    u32 c1 = MakeRGBA(red2, green2, blue2, 255);
    u32 c2, c3;

    if (color1 > color2)
    {
      c2 = MakeRGBA(DXTBlend(red2, red1), DXTBlend(green2, green1), DXTBlend(blue2, blue1), 255);
      c3 = MakeRGBA(DXTBlend(red1, red2), DXTBlend(green1, green2), DXTBlend(blue1, blue2), 255);
    }
    else
    {
      // color[3] is the same as color[2] (average of both colors), but transparent.
      // This differs from DXT1 where color[3] is transparent black.
      c2 = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 255);
      c3 = MakeRGBA((red1 + red2) / 2, (green1 + green2) / 2, (blue1 + blue2) / 2, 0);
    }

    return {c0, c1, c2, c3};
  }

  DOLPHIN_NO_INLINE void Miss(cmprTexelData data, u32 cache_set, u32 cache_tag)
  {
    const u16 color1 = Common::swap16(data.color1);
    const u16 color2 = Common::swap16(data.color2);

    entries[cache_set].colors = ComputeColors(color1, color2);
    entries[cache_set].tag = cache_tag;
  }

public:
  constexpr CMPRColorCache()
  {
    // Since we don't have a dedicated valid bit, and we zero-initialize everything, we need to
    // pre-populate tag 0 (corresponding to colors 0 and 0) with the correct data.
    constexpr u32 zero_set = GetCacheSet(0, 0);
    constexpr u32 zero_tag = GetCacheTag(0, 0);
    static_assert(zero_tag == 0);
    entries[zero_set].tag = zero_tag;
    entries[zero_set].colors = ComputeColors(0, 0);
  }

  u32* LookUp(cmprTexelData data)
  {
    // It's quite expensive to derive the four colors of each block.
    // So we cache then, indexed by the two colors.
    const u32 cache_set = GetCacheSet(data.color1, data.color2);
    const u32 cache_tag = GetCacheTag(data.color1, data.color2);

    if (entries[cache_set].tag != cache_tag)
    {
      Miss(data, cache_set, cache_tag);
    }

    return entries[cache_set].colors.data();
  }

private:
  struct CacheEntry
  {
    u32 tag = {};
    std::array<u32, 4> colors = {};
  };
  std::array<CacheEntry, 1 << CACHE_HASH_BITS> entries = {};
};

static CMPRColorCache s_cmpr_cache;
}  // namespace
```

Reply:
Just to follow up on this: I found a case where this does indeed happen: in Mario Kart Wii, at the end of Toad's Factory, there's a mesh-like texture in a pit (`tex1_256x256_m_dec381016fad805a_14.png`), and it fits the c1 = 0, c2 = 0 case. There probably are a bunch of other examples, but this is the first one I actually found where this ends up being the case.